crates/gen-spidermonkey/src/lib.rs

```diff
@@ -0,0 +1,2180 @@
+///
+/// * Export glue Wasm code (on the Wasm stack)
+///
+///   This is a synthesized function that understands both the canonical ABI and
+///   the SpiderMonkey API. It translates incoming arguments from the canonical
+///   ABI representation into SpiderMonkey values, calls the JS function that
+///   implements this export with those values, and then translates the
+///   function's outgoing results from SpiderMonkey values into the canonical
+///   ABI representation.
+///
+/// * JavaScript function implementing the WITX signature (inside
+///   `spidermonkey.wasm`'s internal stack)
+///
+///   This is the user-written JavaScript function that is being exported. It
+///   accepts and returns the JavaScript values that correspond to the interface
+///   types used in the WITX signature.
+pub struct SpiderMonkeyWasm {
+    i64_return_pointer_area_size: usize,
+
+    num_import_functions: Option<u32>,
+    num_export_functions: Option<u32>,
+
+    import_spidermonkey: bool,
+
+    /// Function types that we use in this Wasm module.
+    types: wasm_encoder::TypeSection,
+
+    /// A map from wasm signature to its index in the `self.types` types
+    /// section. We use this to reuse earlier type definitions when possible.
+    wasm_sig_to_index: HashMap<WasmSignature, u32>,
+
+    /// The imports section containing the raw canonical ABI function imports
+    /// for each imported function we are wrapping in glue.
+    imports: wasm_encoder::ImportSection,
+
+    /// The glue functions we've generated for imported canonical ABI functions
+    /// thus far.
+    import_glue_fns: Vec<wasm_encoder::Function>,
+
+    /// A map from `module_name -> func_name -> (index, num_args)`.
+    import_fn_name_to_index: HashMap<String, HashMap<String, (u32, u32)>>,
+
+    exports: wasm_encoder::ExportSection,
+
+    /// The glue functions we've generated for exported canonical ABI functions
+    /// thus far, and their type index.
+    export_glue_fns: Vec<(wasm_encoder::Function, u32)>,
+
+    data_segments: DataSegments,
+
+    sizes: witx2::SizeAlign,
+}
+
+impl Default for SpiderMonkeyWasm {
+    fn default() -> Self {
+        SpiderMonkeyWasm {
+            i64_return_pointer_area_size: 0,
+            num_import_functions: None,
+            num_export_functions: None,
+            import_spidermonkey: false,
+            types: wasm_encoder::TypeSection::new(),
+            wasm_sig_to_index: Default::default(),
+            imports: wasm_encoder::ImportSection::new(),
+            import_glue_fns: Default::default(),
+            import_fn_name_to_index: Default::default(),
+            exports: wasm_encoder::ExportSection::new(),
+            export_glue_fns: Default::default(),
+            data_segments: DataSegments::new(1),
+            sizes: Default::default(),
+        }
+    }
+}
+
+lazy_static! {
+    /// Functions exported from `spidermonkey.wasm`
+    static ref SMW_EXPORTS: Vec<(&'static str, WasmSignature)> = vec![
```

Comment:
What would the benefit of that be?

This is "static" data, but can't be an actual `static` because `WasmSignature` contains `Vec`s. So if we were to put this inside `SpiderMonkeyWasm` then it would just be dynamically creating copies of the same static data for every `SpiderMonkeyWasm` instance, which seems unnecessary?

Reply:
I've always felt that `lazy_static` is a non-rust-ism since it just leaks everything. I don't really care either way though.